File: Binary Tree Maximum Path Sum.cpp
// Leetcode Problem 124: Binary Tree Maximum Path Sum
// C++ CODE
#include <algorithm>
#include <climits>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int maxSum = INT_MIN;

    int maxGain(TreeNode* node) {
        if (!node) {
            return 0;
        }

        // Recursively calculate max gain from subtrees; clamp negative sums to 0
        int leftGain = std::max(0, maxGain(node->left));
        int rightGain = std::max(0, maxGain(node->right));

        // Max path sum considering the current node as the path root/split point
        int currentPathSum = node->val + leftGain + rightGain;

        // Update the global maximum path sum
        maxSum = std::max(maxSum, currentPathSum);

        // Return maximum single branch path to extend upwards to parent
        return node->val + std::max(leftGain, rightGain);
    }

public:
    int maxPathSum(TreeNode* root) {
        maxGain(root);
        return maxSum;
    }
};

File: Binary Tree Maximum Path Sum.java
// Leetcode Problem 124: Binary Tree Maximum Path Sum
// JAVA CODE

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxSum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        maxGain(root);
        return maxSum;
    }

    private int maxGain(TreeNode node) {
        if (node == null) {
            return 0;
        }

        // Recursively compute the maximum gain from left and right subtrees.
        // Clamped to 0 to ignore paths with negative sums.
        int leftGain = Math.max(0, maxGain(node.left));
        int rightGain = Math.max(0, maxGain(node.right));

        // Price of current node being the root of the path (split point)
        int currentPathSum = node.val + leftGain + rightGain;

        // Update the global maximum path sum
        maxSum = Math.max(maxSum, currentPathSum);

        // Return the maximum branch sum that can be extended to the parent
        return node.val + Math.max(leftGain, rightGain);
    }
}